File: core/src/main/java/com/graphhopper/routing/ch/NodeOrderingProvider.java
package com.graphhopper.routing.ch;

public interface NodeOrderingProvider {

    static NodeOrderingProvider identity(int nodes) {
        return new NodeOrderingProvider() {
            @Override
            public int getNodeIdForLevel(int level) {
                return level;
            }

            @Override
            public int getNumNodes() {
                return nodes;
            }
        };
    }

    static NodeOrderingProvider fromArray(int... nodes) {
        return new NodeOrderingProvider() {
            @Override
            public int getNodeIdForLevel(int level) {
                return nodes[level];
            }

            @Override
            public int getNumNodes() {
                return nodes.length;
            }
        };
    }

    int getNodeIdForLevel(int level);

    int getNumNodes();
}

File: core/src/main/kotlin/com/graphhopper/routing/ch/NodeOrderingProvider.kt
package com.graphhopper.routing.ch

interface NodeOrderingProvider {

    fun getNodeIdForLevel(level: Int): Int

    fun getNumNodes(): Int

    companion object {
        @JvmStatic
        fun identity(nodes: Int): NodeOrderingProvider = object : NodeOrderingProvider {
            override fun getNodeIdForLevel(level: Int): Int = level

            override fun getNumNodes(): Int = nodes
        }

        @JvmStatic
        fun fromArray(vararg nodes: Int): NodeOrderingProvider = object : NodeOrderingProvider {
            override fun getNodeIdForLevel(level: Int): Int = nodes[level]

            override fun getNumNodes(): Int = nodes.size
        }
    }
}
